### DISCUSSION TITLE
brew install - curl: (7) Couldn't connect to server

### DISCUSSION THREAD

**Discussion by scheinercc** [Homebrew/discussions#1036] [Everyday usage] [answered]
Hi there,
I am seeing this error a lot atm
```
$ brew install node@14
==> Downloading https://homebrew.bintray.com/bottles/node%4014-14.16.0_1.catalin
==> Downloading from https://d29vzk4ow07wi7.cloudfront.net/0767567e719f2b2abc841

curl: (7) Couldn't connect to server
Error: Failed to download resource "node@14"
Download failed: https://homebrew.bintray.com/bottles/node%4014-14.16.0_1.catalina.bottle.tar.gz
```

Yesterday after running into this error while trying to `brew upgrade` I eventually restarted my computer.
Re-running `brew upgrade` downloaded one out of 15 packages, before throwing the above error for all others again. I then restarted my terminal and 5 x it downloaded the first package only. After that it got choppy. Sometimes it downloaded one often none again. Eventually I got them all.

Here is one of the errors from yesterday
```
==> Upgrading sphinx-doc 3.4.3 -> 3.5.2
==> Downloading https://homebrew.bintray.com/bottles/sphinx-doc-3.5.2.catalina.bottle.tar.gz

curl: (7) Couldn't connect to server
Error: Failed to download resource "sphinx-doc"
Download failed: https://homebrew.bintray.com/bottles/sphinx-doc-3.5.2.catalina.bottle.tar.gz
```

Today I tried the above `install`.

Trying the links like https://homebrew.bintray.com/bottles/node%4014-14.16.0_1.catalina.bottle.tar.gz in the browser works fine.

This is on macOS 10.15.7.
I tried it on several different networks.

Any idea what might be going on?

Also having this issue with `brew doctor` right now. Not sure if this is related, though.
https://stackoverflow.com/questions/66591643/how-to-fix-brew-doctor-output-with-double-question-marks

**Comment by carlocab** [comment]
What's the output of `brew doctor` and `brew config`?

**Reply by scheinercc** [reply]
My `brew doctor` issue linked above got fixed thanks to Theoretical Economist over at SO.

The server error still persists.

Now I am only left with
```
$ brew doctor
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Some installed formulae are deprecated or disabled.
You should find replacements for the following formulae:
  spdylay
```

Here is the full config:
```
brew config
HOMEBREW_VERSION: 3.0.5-18-g433cd83
ORIGIN: https://github.com/Homebrew/brew
HEAD: 433cd83cbc186afbe970289557ceee3f33c7383a
Last commit: 9 hours ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 77474e6293025aeea3a7e25e86883967eaa02675
Core tap last commit: 3 hours ago
Core tap branch: master
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 8
Homebrew Ruby: 2.6.3 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.3_2/bin/ruby
CPU: octa-core 64-bit skylake
Clang: 12.0 build 1200
Git: 2.24.3 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 7.64.1 => /usr/bin/curl
macOS: 10.15.7-x86_64
CLT: 12.4.0.0.1.1610135815
Xcode: N/A
```

**Reply by carlocab** [reply]
Hmm. Do you have Homebrew `curl` installed? If you do, try
```
HOMEBREW_FORCE_BREWED_CURL=1 brew install node@14
```

**Reply by scheinercc** [reply]
```
$ HOMEBREW_FORCE_BREWED_CURL=1 brew install node@14
Updating Homebrew...
==> Auto-updated Homebrew!
Updated 2 taps (homebrew/core and homebrew/cask).
==> Updated Formulae
Updated 2 formulae.
==> Updated Casks
Updated 2 casks.

==> Downloading https://homebrew.bintray.com/bottles/node%4014-14.16.0_1.catalina.bottle.tar.gz
==> Downloading from https://d29vzk4ow07wi7.cloudfront.net/0767567e719f2b2abc8413a2f3f4e546ad1f5a1433ae32932976a8ecc8f21bce?response-content-disposition=attachment%3Bfilename%3D%22node%4014-14.16.0_1.catalina.bot
curl: (7) Couldn't connect to server

Error: Failed to download resource "node@14"
Download failed: https://homebrew.bintray.com/bottles/node%4014-14.16.0_1.catalina.bottle.tar.gz
```

**Reply by scheinercc** [reply]
```
curl -I https://d29vzk4ow07wi7.cloudfront.net/0767567e719f2b2abc8413a2f3f4e546ad1f5a1433ae32932976a8ecc8f21bce?response-content-disposition=attachment%3Bfilename%3D%22node%4014-14.16.0_1.catalina.bot
```

returns
```
HTTP/1.1 403 Forbidden
Server: CloudFront
Date: Fri, 12 Mar 2021 03:28:22 GMT
Content-Type: text/xml
Content-Length: 146
Connection: keep-alive
X-Cache: Error from cloudfront
Via: 1.1 6785379936d15b44a779e5f13a6567df.cloudfront.net (CloudFront)
X-Amz-Cf-Pop: SFO5-C3
X-Amz-Cf-Id: Vw0FiBosLmhgGSKTM_YEwCJxnMGzB060KdLAOVH8Hk3MAOZ7PskkRQ==
```

**Reply by scheinercc** [reply]
FYI, also tried via VPN without any luck.

**Reply by gromgit** [reply]
What does the following output?
```
curl -I https://homebrew.bintray.com/bottles/node%4014-14.16.0_1.catalina.bottle.tar.gz
curl -I https://homebrew.bintray.com/bottles/node-15.11.0.catalina.bottle.tar.gz
```

**Reply by scheinercc** [reply]
```
$ curl -I https://homebrew.bintray.com/bottles/node%4014-14.16.0_1.catalina.bottle.tar.gz
HTTP/1.1 200 OK
Server: nginx
Date: Mon, 15 Mar 2021 03:49:08 GMT
Content-Type: application/gzip
Content-Length: 17072401
Connection: keep-alive
Last-Modified: Sat, 27 Feb 2021 23:51:39 GMT
Accept-Ranges: none
ETag: 0767567e719f2b2abc8413a2f3f4e546ad1f5a1433ae32932976a8ecc8f21bce
X-Checksum-Sha1: 3a4d78c0f7fc588bacc113ee5b87b4865dfe4d4a
X-Checksum-Sha2: 0767567e719f2b2abc8413a2f3f4e546ad1f5a1433ae32932976a8ecc8f21bce
```

```
curl -I https://homebrew.bintray.com/bottles/node-15.11.0.catalina.bottle.tar.gz
HTTP/1.1 200 OK
Server: nginx
Date: Mon, 15 Mar 2021 03:49:35 GMT
Content-Type: application/gzip
Content-Length: 16213067
Connection: keep-alive
Last-Modified: Wed, 03 Mar 2021 13:19:40 GMT
Accept-Ranges: none
ETag: 1dc4ea77c4196c8376eb6d81401f7bdb357c49ce456780659a2a4e01685cea4e
X-Checksum-Sha1: 091ff7e6d6c28db470450cb8adf130ff21e30f27
X-Checksum-Sha2: 1dc4ea77c4196c8376eb6d81401f7bdb357c49ce456780659a2a4e01685cea4e
```

**Reply by gromgit** [reply]
OK, now run:
```
curl -o /tmp/node%4014-14.16.0_1.catalina.bottle.tar.gz -L --trace-ascii /tmp/curl.dl-node.txt https://homebrew.bintray.com/bottles/node%4014-14.16.0_1.catalina.bottle.tar.gz
```
then post the contents of `/tmp/curl.dl-node.txt` as a gist if the command fails.

**Reply by scheinercc** [reply]
that finished without an error
```
$ curl -o /tmp/node%4014-14.16.0_1.catalina.bottle.tar.gz -L --trace-ascii /tmp/curl.dl-node.txt https://homebrew.bintray.com/bottles/node%4014-14.16.0_1.catalina.bottle.tar.gz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100 16.2M  100 16.2M    0     0  2832k      0  0:00:05  0:00:05 --:--:-- 3617k
```

Here is the Gist: https://gist.github.com/scheinercc/f968ad90ab20453dc7a2c2ef237a357a

**Reply by gromgit** [reply]
Interesting, `curl` by itself picks up and processes the Cloudfront redirect correctly. How about:
```
HOMEBREW_CURL_VERBOSE=1 brew install node@14
```

**Reply by scheinercc** [reply]
```
$ HOMEBREW_CURL_VERBOSE=1 brew install node@14
==> Downloading https://homebrew.bintray.com/bottles/node%4014-14.16.0_1.catalina.bottle.tar.gz
==> Downloading from https://d29vzk4ow07wi7.cloudfront.net/0767567e719f2b2abc8413a2f3f4e546ad1f5a1433ae329329
*   Trying 13.225.143.12...
* TCP_NODELAY set
* Immediate connect fail for 13.225.143.12: Unknown error: -536870212
*   Trying 13.225.143.103...
* TCP_NODELAY set
* Immediate connect fail for 13.225.143.103: Unknown error: -536870212
*   Trying 13.225.143.82...
* TCP_NODELAY set
* Immediate connect fail for 13.225.143.82: Unknown error: -536870212
*   Trying 13.225.143.223...
* TCP_NODELAY set
* Immediate connect fail for 13.225.143.223: Unknown error: -536870212
* Closing connection 0

curl: (7) Couldn't connect to server
Error: Failed to download resource "node@14"
Download failed: https://homebrew.bintray.com/bottles/node%4014-14.16.0_1.catalina.bottle.tar.gz
```

I had a quick search for "Unknown error: -536870212", but nothing specific came up. I gave logging out of Apple ID, restarting and logging back in another shot. Still the same result.

**Reply by carlocab** [reply]
If you still have `/tmp/node%4014-14.16.0_1.catalina.bottle.tar.gz` there (or you can just download it again), you can try
```
brew install /tmp/node%4014-14.16.0_1.catalina.bottle.tar.gz
```
In theory, you should be able to install directly from the bottle tarball, but I should mention that I often get weird issues when I do that. YMMV.

**Reply by gromgit** [reply]
How about:
```
type -p curl
HOMEBREW_CURL_VERBOSE=1 brew -v -d install node@14
```

**Reply by scheinercc** [reply]
Thanks carlocab and gromgit for all your efforts so far. That works for the example of node@14. I will keep a bookmark to this install/upgrade option. If anyone comes up with another idea for me to get to the bottom of the connection error for all bottles atm, would be much appreciated.

**Reply by scheinercc** [reply]
```
$ type -p curl
curl is /usr/local/opt/curl/bin/curl
```

```
$ HOMEBREW_CURL_VERBOSE=1 brew -v -d install node@14
Error: Unknown command: -d
```

**Reply by gromgit** [reply]
Sorry, mistyped:
```
HOMEBREW_CURL_VERBOSE=1 brew install -v -d node@14
```

**Reply by scheinercc** [reply]
Sorry I didn't pick it up. I uninstalled `node@14` again before running the command. 
Here is the output: https://gist.github.com/scheinercc/5a7f29855fa16723923a9bf28bbd5178

**Reply by gromgit** [reply]
OK, that gist says:
1. `brew` is using your system `curl`
2. `brew install node@14` now works

If you want to be doubly-sure:
```
brew uninstall node@14
brew install node@14
```
to see if you can trigger your original error, though I'm starting to suspect a transient Cloudfront CDN issue that's been quietly resolved.

**Reply by scheinercc** [reply]
It is weird. I still do get this error - just tried it with a node version I am sure I have not installed*:
```
brew install node@10
==> Downloading https://homebrew.bintray.com/bottles/node%4010-10.24.0_1.catalina.bottle.tar.gz
==> Downloading from https://d29vzk4ow07wi7.cloudfront.net/efd6439e5f6afc43b8d84e98a61789b73f8e5ec6c9f6909668255e369a7589b5?response-content-disposition=attachment%3Bfilename%3D%22node%4010-10.24.0_1.catalina.bottle.tar.gz

curl: (7) Couldn't connect to server
Error: Failed to download resource "node@10"
Download failed: https://homebrew.bintray.com/bottles/node%4010-10.24.0_1.catalina.bottle.tar.gz
```

*v14 worked as it found it in Cache "Already downloaded"
```
==> Downloading https://homebrew.bintray.com/bottles/node%4014-14.16.0_1.catalina.bottle.tar.gz
Already downloaded: /Users/albertscheiner/Library/Caches/Homebrew/downloads/5d00157fd4c3b84fe6ddcdd5870e89b9e94ff3a77cfaa0fbff0cc3fa647ae301--node@14-14.16.0_1.catalina.bottle.tar.gz
==> Pouring node@14-14.16.0_1.catalina.bottle.tar.gz
...
```

**Reply by gromgit** [reply]
That's...weird. Can you move `/Users/albertscheiner/Library/Caches/Homebrew/downloads/5d00157fd4c3b84fe6ddcdd5870e89b9e94ff3a77cfaa0fbff0cc3fa647ae301--node@14-14.16.0_1.catalina.bottle.tar.gz` to a temp directory and try `brew install node@14` again?

And if that fails, see if `HOMEBREW_CURL_VERBOSE=1 brew install -v -d node@14` works. If it does...I'm stumped.

**Reply by scheinercc** [reply]
removed the temp file, uninstalled `node@14` and the default command fails again:
```
$ brew install node@14
==> Downloading https://homebrew.bintray.com/bottles/node%4014-14.16.0_1.catalina.bottle.tar.gz
==> Downloading from https://d29vzk4ow07wi7.cloudfront.net/0767567e719f2b2abc8413a2f3f4e546ad1f5a1433ae32932976a8ecc8f21bce?response-content-disposition=attachment%3Bfilename%3D%22node%4014-14.16.0_1.catalina.bottle.tar.gz

curl: (7) Couldn't connect to server
Error: Failed to download resource "node@14"
Download failed: https://homebrew.bintray.com/bottles/node%4014-14.16.0_1.catalina.bottle.tar.gz
```

and verbose works like a charm 🤔 https://gist.github.com/scheinercc/7c7643b58a11b13c682baba9ffc38e0d

**Reply by carlocab** [reply]
Are you able to replicate that consistently? i.e. `brew install` not working, but `HOMEBREW_CURL_VERBOSE=1 brew install -v -d` does?

**Reply by scheinercc** [reply]
~Yeap. So far 100%~

**_Update:_**
Nope. It was true for `node@14` and my last test `node@10`, but now I randomly tried to install `align`, which as expected failed for `brew install align`, but also failed for `HOMEBREW_CURL_VERBOSE=1 brew install -v -d align` [see Gist for details](https://gist.github.com/scheinercc/1ff5f799478acd775f8b1eb7959a0532)

Requesting `https://homebrew.bintray.com/bottles/align-1.7.5.catalina.bottle.tar.gz` in the browser works.

**_Update 2:_**
I tried two more, `apollo-cli` and `zzz`
1. `HOMEBREW_CURL_VERBOSE=1 brew install -v -d apollo-cli` => worked
2. `brew uninstall apollo-cli`+ rm cached file
3. `brew install apollo-cli` => failed
4. `HOMEBREW_CURL_VERBOSE=1 brew install -v -d apollo-cli` => worked
5. steps 2 - 4 again with the same result
6. `brew install zzz` => failed
7. `HOMEBREW_CURL_VERBOSE=1 brew install -v -d zzz` => failed [see Gist for details](https://gist.github.com/scheinercc/87d8909cab7ee87c98d44ca2d3eccd6f)

I was able to replicate it consistently with `node@14`, `node@10` and `apollo-cli`, but `align` and `zzz` failed.

**Reply by gromgit** [reply]
Do you have any firewall or anti-virus program running on your machine, or somewhere in your network?

**Reply by scheinercc** [reply]
I was sure I hadn't, but I forgot about Lulu: https://github.com/objective-see/LuLu 😞 disabling it fixed it. Thanks for all your help guys!

**Comment by scheinercc** [comment, marked as answer]
Issue was not homebrew related. Fixed by disabling firewall (needs rules adjusted). Thanks to carlocab & gromgit for all the help.

**Comment by timotheecour** [comment]
I'm seeing https://dev.azure.com/nim-lang/Nim/_build/results?buildId=14320&view=logs&jobId=30931762-47c4-53b3-6a83-316eb5a6b9d7&j=30931762-47c4-53b3-6a83-316eb5a6b9d7&t=674d5251-e236-506b-826a-9bb8444a0855

when running `brew install boehmgc make sfml` in CI, could that be related?
it only started today IIRC

curl: (35) LibreSSL SSL_connect: SSL_ERROR_SYSCALL in connection to homebrew.bintray.com:443 
Error: Failed to download resource "make"

(see https://github.com/timotheecour/Nim/issues/653 for similar errors, but with different packages, all since today)

```
Starting: Install dependencies (OSX)
==============================================================================
Task         : Bash
Description  : Run a Bash script on macOS, Linux, or Windows
Version      : 3.182.0
Author       : Microsoft Corporation
Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/utility/bash
==============================================================================
Generating script.
Script contents:
brew install boehmgc make sfml
========================== Starting Command Output ===========================
/bin/bash --noprofile --norc /Users/runner/work/_temp/52fe84b9-e97e-4df2-951d-9bbe32c78c2e.sh
Warning: bdw-gc 8.0.4_2 is already installed and up-to-date.
To reinstall 8.0.4_2, run:
  brew reinstall bdw-gc
==> Downloading https://homebrew.bintray.com/bottles/make-4.3.catalina.bottle.1.tar.gz
curl: (35) LibreSSL SSL_connect: SSL_ERROR_SYSCALL in connection to homebrew.bintray.com:443 
Error: Failed to download resource "make"
Download failed: https://homebrew.bintray.com/bottles/make-4.3.catalina.bottle.1.tar.gz
##[error]Bash exited with code '1'.
Finishing: Install dependencies (OSX)
```

**Reply by scheinercc** [reply]
I am pretty certain it was one of my local firewall rules causing the issue. `brew install` works for me again.

**Reply by carlocab** [reply]
We're seeing this in Homebrew CI too. Might be a Bintray issue.

**Reply by cybergopnik** [reply]
In my case it was LuLu's rule mistakenly writen by me too.
